reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/utils/serializer/SerializerImpl.kt

```diff
@@ -1,54 +1,68 @@
 package com.badoo.reaktive.utils.serializer
 
-import com.badoo.reaktive.utils.Lock
 import com.badoo.reaktive.utils.queue.Queue
-import com.badoo.reaktive.utils.queue.isNotEmpty
-import com.badoo.reaktive.utils.queue.take
-import com.badoo.reaktive.utils.synchronized
+import com.badoo.reaktive.utils.queue.isEmpty
 
 internal abstract class SerializerImpl<in T>(queue: Queue<T>) : Serializer<T> {
 
     private var queue: Queue<T>? = queue
-    private val lock = Lock()
+    private val monitor = Any()
     private var isDraining = false
 
     override fun accept(value: T) {
-        lock
-            .synchronized {
-                queue
-                    ?.apply { offer(value) }
-                    ?.takeUnless { isDraining }
-                    ?.also { isDraining = true }
+        val q =
+            synchronized(monitor) {
+                val q = queue ?: return
+
+                if (isDraining) {
+                    q.offer(value)
+                    return
+                }
+
+                isDraining = true
+                q
             }
-            ?.drain()
+
+        if (!processValue(value)) {
+            return
+        }
+
+        q.drain()
     }
 
     override fun clear() {
-        lock.synchronized {
+        synchronized(monitor) {
             queue?.clear()
         }
     }
 
     private fun Queue<T>.drain() {
         while (true) {
-            lock
-                .synchronized {
-                    if (isNotEmpty) {
-                        take()
-                    } else {
+            val value =
+                synchronized(monitor) {
+                    if (isEmpty) {
                         onDrainFinished(false)
                         return
                     }
+
+                    poll()!!
```

Comment:
Why `poll()!!` instead of `take()`?

Reply:
Because it checks isEmpty under the hood and I would like to avoid double check.
Discovered NPE and fixed.
Removed `take` as not being used.